### PR TITLE
WIP: packaging, installation, and testing

### PR DESCRIPTION
**Pull request by ceball**
* dependencies in one place (setup.py)
* added noarch:python conda package, with metadata read from setup.py
* when a tag is pushed, will build, install, test, and - if tests successful - upload packages.
* conda package uploaded to anaconda.org pyviz when release style tag pushed (label/dev for pre-release style tag)
* travis will normally just test a 'dev environment'
* adds caching on travis for dev tests (the ones that run every time). Will have to see what impact this has on earthsim build times (haven't yet tested here). Caching is disabled for packaging step - wouldn't have much effect anyway (conda build creates fresh test environments).
* adds testing on windows (appveyor)
* for testing notebooks, use nbsmoke (check for flakes, check they run without error)
* added earthsim command for installing examples and downloading data (`earthsim install_examples`, `earthsim download_data`); see https://github.com/pyviz/pvutil/blob/master/README.md
* test commands all defined in tox.ini (and that's the only place they're defined).
* doit/pyct used to document tasks in a runnable way (runnable locally, on CI, across platforms). Work in progress. But all tasks just run standard tools, e.g. develop install is "conda install dependencies, then `pip install -e . --no-deps`. Packaging uses conda build. Testing uses pytest (or pytest plugins). See see https://github.com/pyviz/pyct/blob/master/README.md for a draft overview of pyct.

To do:
* [x] earthsim conda package is now waiting on updated geoviews conda package (https://github.com/ioam/geoviews/pull/138) and quest conda package (https://github.com/erdc/quest/issues/5)?
* [ ] can we simplify list of channels after some releases of other packages?
* [x] (edit: using holoviews examples packaging mechanism) decide if `examples` symlink to `earthsim/examples` is ok. Negative: windows git developers may need to set up their git/system to support it. Negative: you may not like "seeing notebooks twice when you are operating in git repo". Positive: no code needed to copy files at package time. Positive: no code needed to find examples that depends on whether development or package install.

To do for CB:
* [ ] update pins for dependencies
* [ ] grep for TODOs and check they're done
* [ ] update docs for installation

Subsequent PR after this one is merged should enable flake checks and fix flakes (many unused imports...): #56.

### CODE REVIEW

File: .appveyor.yml [hunk 1]
```diff
@@ -0,0 +1,20 @@
+environment:
+  global:
+    CHANS_DEV: "-c pyviz/label/dev -c ioam -c bokeh -c conda-forge -c erdc -c defaults"
```
Comment (jbednar) [inline .appveyor.yml]: That is quite a list of channels!  Do we really need all of them?

File: .travis.yml [hunk 2]
```diff
@@ -5,53 +5,87 @@ stages:
...
-#        - brew uninstall --ignore-dependencies geos gdal postgis
-#      if: branch = master AND type != pull_request
+#      stage: doc
+#      script:
+#        - doit install_doc_dependencies
+#        - doit download_sample_data
+#        - travis_wait 60 doit docs
+#      deploy:
+#        edge:
+#          branch: v1.8.47
```
Comment (jbednar) [inline .travis.yml]: Where does the string "v1.8.47" come from?

File: setup.py [hunk 3]
```diff
@@ -1,12 +1,131 @@
...
+########## dependencies ##########
+
+# TODO: need to sort through these, and what's the split between
+# required/recommended? I made a fairly random cut...
+
+install_requires = [
+    'bokeh',
+    'datashader',
+    'geoviews',  # TODO > ?
+    'holoviews', # TODO > ?
```
Comment (jbednar) [inline setup.py]: What's to do?  Pin the version?

File: setup.py [hunk 4]
```diff
@@ -1,12 +1,131 @@
...
+    'jupyter',
+    'numpy',
+    'param >=1.6.1',
+    'parambokeh',
+    'paramnb',
```
Comment (jbednar) [inline setup.py]: Hopefully soon to be deleted?

File: setup.py [hunk 5]
```diff
@@ -1,12 +1,131 @@
...
+    'pony',
+    'scikit-image',
+    'go-spatial',
+    'descartes',
+    'gsshapy',
+    'gssha',
+    'cartopy',
+    'xarray',
+    'filigree',
+    'pvutil',
```
Comment (jbednar) [inline setup.py]: Some of the packages in this list (at least xarray, cartopy, matplotlib, gdal) are already dependencies of the ones in `install_requires`, and thus should only be listed here if they are explicitly being imported in one of the examples.

Comment (ceball) [reply setup.py]: Thanks. Yes, still a TODO to sort through these.

File: .appveyor.yml (hunk 1 above)
Comment (ceball) [reply .appveyor.yml]: I guess I'm expecting it to become something like `-c pyviz/label/dev -c erdc -c conda-forge` for developers, and `-c pyviz -c erdc -c conda-forge` for released packages. And then as we get more onto c-f and defaults, maybe releases will eventually not need extra channels (someone can just use defaults or c-f).

File: .travis.yml (hunk 2 above)
Comment (jbednar) [reply .travis.yml]: That should be a comment in the file, not here in the PR, right? It stands out, and we'll want to be able to remove it with confidence later when it's fixed.

**Comment by ceball** [comment]
Currently seeing this on linux (but not mac or win):

```
E                   ------------------
E                   import os
E                   import numpy as np
E                   import holoviews as hv
E                   import geoviews as gv
E                   import cartopy.crs as ccrs
E                   
E                   from holoviews.streams import PointDraw, PolyEdit, BoxEdit, PolyDraw
E                   
E                   url = 'http://c.tile.openstreetmap.org/{Z}/{X}/{Y}.png'
E                   tiles = gv.WMTS(url, extents=(-91, 32.2, -90.8, 32.4), crs=ccrs.PlateCarree())
E                   
E                   hv.extension('bokeh')
E                   ------------------
E                   
E                   ---------------------------------------------------------------------------
E                   ImportError                               Traceback (most recent call last)
E                   <ipython-input-1-a795f3aeb311> in <module>()
E                         2 import numpy as np
E                         3 import holoviews as hv
E                   ----> 4 import geoviews as gv
E                         5 import cartopy.crs as ccrs
E                         6 
E                   
E                   ~/miniconda/envs/test-environment/lib/python3.6/site-packages/geoviews/__init__.py in <module>()
E                         6                       QuadMesh, VectorField, HexTiles, Labels)
E                         7 from . import data                                  # noqa (API import)
E                   m----> 8 from . import operation                             # noqa (API import)
E                         9 from . import plotting                              # noqa (API import)
E                        10 from . import feature                               # noqa (API import)
E                   
E                   ~/miniconda/envs/test-environment/lib/python3.6/site-packages/geoviews/operation/__init__.py in <module>()
E                         5 from .. import element as gv_element
E                         6 from ..element import _Element
E                   ----> 7 from .projection import ( # noqa (API import)
E                         8     project_image, project_path, project_shape, project_points,
E                         9     project_graph, project_quadmesh, project
E                   
E                   ~/miniconda/envs/test-environment/lib/python3.6/site-packages/geoviews/operation/projection.py in <module>()
E                         3 
E                         4 from cartopy import crs as ccrs
E                   ----> 5 from cartopy.img_transform import warp_array, _determine_bounds
E                         6 from holoviews.core.util import cartesian_product, get_param_values
E                         7 from holoviews34m.operation import Operation
E                   
E                   ~/miniconda/envs/test-environment/lib/python3.6/site-packages/cartopy/img_transform.py in <module>()
E                        24 
E                        25 import numpy as np
E                   ---> 26 import scipy.spatial
E                        27 
E                        28 import cartopy.crs as ccrs
E                   
E                   ~/miniconda/envs/test-environment/lib/python3.6/site-packages/scipy/spatial/__init__.py in <module>()
E                        91 from __future__ import division, print_function, absolute_import
E                        92 
E                   ---> 93 from .kdtree import *
E                        94 from .ckdtree import *
E                        95 from .qhull import *
E                   
E                   ~/miniconda/envs/test-environment/lib/python3.6/site-packages/scipy/spatial/kdtree.py in <module>()
E                         6 import numpy as np
E                         7 from heapq import heappush, heappop
E                   ----> 8 import scipy.sparse
E                         9 
E                        10 __all__ = ['minkowski_distance_p', 'minkowski_distance',
E                   
E                   ~/miniconda/envs/test-environment/lib/python3.6/site-packages/scipy/sparse/__init__.py in <module>()
E                       227 
E                       228 from .base import *
E                   --> 229 from .csr import *
E                       230 from .csc import *
E                       231 from .lil import *
E                   
E                   ~/miniconda/envs/test-environment/lib/python3.6/site-packages/scipy/sparse/csr.py in <module>()
E                        13 from .base import spmatrix
E                        14 
E                   ---> 15 from ._sparsetools import csr_tocsc, csr_tobsr, csr_count_blocks, \
E                        16         get_csr_submatrix, csr_sample_values
E                        17 from .sputils import (upcast, isintlike, IndexMixin, issequence,
E                   
E                   ImportError: /home/travis/miniconda/envs/test-environment/lib/python3.6/site-packages/zmq/backend/cython/../../../../.././libstdc++.so.6: version `CXXABI_1.3.9' not found (required by /home/travis/miniconda/envs/test-environment/lib/python3.6/site-packages/scipy/sparse/_sparsetools.cpython-36m-x86_64-linux-gnu.so)
```

**Pull request closed without merging**